### PR TITLE
[PT FE] Support converting 16bit models by patching

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Add patch function to make 16bit models traceable. This reduce memory consumption and time spent during conversion of models that are natively saved in 16bit precision*

### Tickets:
 - *CVS-136828*

### CODE REVIEW

File: src/frontends/pytorch/src/op/linear.cpp [hunk 1]
```diff
@@ -15,9 +16,23 @@ OutputVector translate_linear(const NodeContext& context) {
...
     num_inputs_check(context, 2, 3);
     auto x = context.get_input(0);
-    auto y = context.get_input(1);
-    auto matmul = context.mark_node(std::make_shared<ov::op::v0::MatMul>(x, y, false, true));
-    return {context.mark_output(make_optional_bias(matmul, context, 2))};
+    auto weight = context.get_input(1);
+    if (weight.get_element_type() == element::f16 || weight.get_element_type() == element::bf16) {
+        // In case of patched linear it can have mixed fp16/bf16 and fp32 input type.
+        // In other cases these conversion is not required.
+        weight = context.mark_node(std::make_shared<ov::op::v1::ConvertLike>(weight, x));
```
Comment (rkazants) [inline src/frontends/pytorch/src/op/linear.cpp:23]: can we use `Convert` if the required type is known?

File: tests/layer_tests/py_frontend_tests/test_torch_frontend.py [hunk 2]
```diff
@@ -681,3 +681,40 @@ def test_output_tuple_names():
...
+    # the approach with patching only works for node with no grad
+    with torch.no_grad():
+        converted_model = convert_model(model_fp16, example_input=example)
+    assert converted_model
+    model_bf16 = ModelWithLinear().bfloat16()
+    patch_model.__make_16bit_traceable(model_bf16)
+    # the approach with patching only works for node with no grad
+    with torch.no_grad():
+        converted_model = convert_model(model_bf16, example_input=example)
+    assert converted_model
```
Comment (rkazants) [inline tests/layer_tests/py_frontend_tests/test_torch_frontend.py:720]: Should we test inference results here?

File: src/frontends/pytorch/src/op/linear.cpp (hunk 1 above)
Comment (mvafin) [reply src/frontends/pytorch/src/op/linear.cpp:23]: `x` type may be still not known, so I don't think that is a good idea

File: src/bindings/python/src/openvino/frontend/pytorch/patch_model.py [hunk 3]
```diff
@@ -79,3 +80,19 @@ def unpatch_model(model, orig_forward_name):
                 print('[ WARNING ] Exception raised during model unpatching. Depending on the exact issue it may lead to broken original model.')
                 print('Original exception details:')
                 print(error)
+
+
+def __make_16bit_traceable(model: torch.nn.Module):
```
Comment (rkazants) [inline src/bindings/python/src/openvino/frontend/pytorch/patch_model.py:85]: Should user patch a model manually? No automatic detection mechanism?
Also, will it be internal or external function? If external, do we need `__` in the method name?

File: src/frontends/pytorch/src/op/linear.cpp (hunk 1 above)
Comment (rkazants) [reply src/frontends/pytorch/src/op/linear.cpp:23]: in TF FE we have helper that check if it is static or dynamic. Based on it, it uses either `Convert` or `ConvertLike`

File: src/bindings/python/src/openvino/frontend/pytorch/patch_model.py (hunk 3 above)
Comment (mvafin) [reply src/bindings/python/src/openvino/frontend/pytorch/patch_model.py:85]: We decided to make it internal for now

**Review by eaidova — APPROVED** [review]

File: tests/layer_tests/py_frontend_tests/test_torch_frontend.py (hunk 2 above)
Comment (mvafin) [reply tests/layer_tests/py_frontend_tests/test_torch_frontend.py:729]: Done

**Review by rkazants — APPROVED** [review]

File: tests/layer_tests/py_frontend_tests/test_torch_frontend.py [hunk 4]
```diff
@@ -681,3 +681,53 @@ def test_output_tuple_names():
...
+
+    patch_model.__make_16bit_traceable(model_fp16)
+    # the approach with patching only works for node with no grad
+    with torch.no_grad():
+        converted_model = convert_model(model_fp16, example_input=example)
+    assert converted_model
+    cm_fp16 = compile_model(converted_model, "CPU")
+    res_fp16 = cm_fp16([x.numpy() for x in example])
+    np.testing.assert_allclose(res_fp16[0], res_ref[0].numpy(), atol=1e-3)
+    np.testing.assert_allclose(res_fp16[1], res_ref[1].numpy(), atol=1e-3)
```
Comment (mvafin) [inline tests/layer_tests/py_frontend_tests/test_torch_frontend.py:722]:
```suggestion
    np.testing.assert_allclose(res_fp16[0], res_ref[0].numpy(), atol=1e-2)
    np.testing.assert_allclose(res_fp16[1], res_ref[1].numpy(), atol=1e-2)
```

**Pull request merged**